In [8]:
import random

In [9]:
def c_look(requests, head):
    """
    C-LOOK disk scheduling algorithm.
    :param requests: List of cylinder requests
    :param head: Current head position
    :return: Total seek count
    """
    size = len(requests)
    distance = 0
    cur_track = head

    # Sort the request list
    requests.sort()

    # Start processing the requests
    for i in range(size):
        next_track = requests[i]
        # Calculate the distance
        distance += abs(cur_track - next_track)
        # Update the current head position
        cur_track = next_track

    # Jumping back to the first request in the direction
    if size != 0:
        distance += abs(requests[0] - cur_track)
        cur_track = requests[0]

    return distance

In [10]:
# Initialize variables
disk_size = 200
initial_head_position = 50

In [11]:
# Different sets of requests
request_sizes = [10, 20, 50, 100]

In [12]:
# Dictionary to store results
results = {}

In [13]:
# Run C-LOOK algorithm for different sizes of requests
for size in request_sizes:
    total_seek = 0
    worst_seek = 0
    for _ in range(1000): # Run 1000 simulations for each set of requests
        requests = random.sample(range(disk_size), size)
        seek = c_look(requests, initial_head_position)
        total_seek += seek
        worst_seek = max(worst_seek, seek)

    # Calculate average and worst-case seek times
    avg_seek = total_seek / 1000
    results[size] = (avg_seek, worst_seek)

In [14]:
results

{10: (363.714, 448),
 20: (404.916, 448),
 50: (433.59, 448),
 100: (443.442, 448)}

Analysis

The average seek time increases with the number of requests. This is expected as more requests generally mean more movement for the disk head.

The worst-case seek time remains constant at 448 cylinders for all sets of requests. This is because the worst-case scenario in C-LOOK is determined by the furthest request from the current head position, and since the disk size and head position are constant, the maximum possible seek distance doesn't change.

The C-LOOK algorithm performs efficiently in terms of reducing the total back-and-forth movement of the disk head, as seen by the relatively low average seek times even for larger sets of requests. This makes it suitable for systems where minimizing the disk arm movement is crucial for performance.